### PR TITLE
Add support for creating repos from template repos

### PR DESCRIPTION
**Pull request by shanempope**
Add API for `POST /repos/:template_owner/:template_repo/generate` endpoint which creates a repository from a template repository.
Developer Blog: https://developer.github.com/changes/2019-07-16-repository-templates-api/
API Reference: https://developer.github.com/v3/repos/#create-repository-using-a-repository-template

Issue: https://github.com/google/go-github/issues/1235

### CODE REVIEW

**Comment by shanempope** [comment]
I signed it!

**Comment by shanempope** [comment]
> Note: Creating and using repository templates is currently available for developers to preview. To access this new endpoint during the preview period, you must provide a custom media type in the Accept header:
> 
> `application/vnd.github.baptiste-preview+json`

And

> Warning: The API may change without advance notice during the preview period. Preview features are not supported for production use. If you experience any issues, contact GitHub Support.

File: github/repos.go [hunk 1]
```diff
@@ -334,6 +338,36 @@ func (s *RepositoriesService) Create(ctx context.Context, org string, repo *Repo
 	return r, resp, nil
 }
 
+// TemplateRepoRequest represents a request to create a repo from a template
```
Comment (gmlewis) [inline github/repos.go]: For the auto-generated godocs, please end the comment in a period.

File: github/repos.go [hunk 2]
```diff
@@ -334,6 +338,36 @@ func (s *RepositoriesService) Create(ctx context.Context, org string, repo *Repo
...
+type TemplateRepoRequest struct {
+	// Name is required when creating a repo.
+	Name        *string `json:"name,omitempty"`
+	Owner       *string `json:"owner,omitempty"`
+	Description *string `json:"description,omitempty"`
+
+	Private *bool `json:"private,omitempty"`
+}
+
+// Generate a repository using a repository template
```
Comment (gmlewis) [inline github/repos.go]: Hmmm... I see that GitHub named this endpoint `generate`, which seems odd to me.

@gauntface - what do you think about calling this endpoint `repositories.CreateFromTemplate` (or maybe just `FromTemplate`?) as that seems to more appropriate to me?

Either way, @shanempope, we will want to make this comment a complete sentence for the auto-generated godocs.  Hopefully something like:

`// CreateFromTemplate generates a repository from a template.`

File: github/repos.go [hunk 3]
```diff
@@ -334,6 +338,36 @@ func (s *RepositoriesService) Create(ctx context.Context, org string, repo *Repo
...
+//
+// GitHub API docs: https://developer.github.com/v3/repos/#create-repository-using-a-repository-template
+func (s *RepositoriesService) Generate(ctx context.Context, templateOwner string, templateRepo string, templateRepoReq *TemplateRepoRequest) (*Repository, *Response, error) {
```
Comment (gmlewis) [inline github/repos.go]: `templateOwner string, templateRepo string,` should be `templateOwner, templateRepo string,`

File: github/repos.go [hunk 4]
```diff
@@ -334,6 +338,36 @@ func (s *RepositoriesService) Create(ctx context.Context, org string, repo *Repo
...
+	u := fmt.Sprintf("repos/%v/%v/generate", templateOwner, templateRepo)
+
+	req, err := s.client.NewRequest("POST", u, templateRepoReq)
+	if err != nil {
+		return nil, nil, err
+	}
+
```
Comment (gmlewis) [inline github/repos.go]: Please add the required preview header here.

File: github/repos_test.go [hunk 5]
```diff
@@ -333,6 +333,66 @@ func TestRepositoriesService_Create_org(t *testing.T) {
...
+
+	templateRepoReq := &TemplateRepoRequest{
+		Name: String("n"),
+	}
+
+	mux.HandleFunc("/repos/to/tr/generate", func(w http.ResponseWriter, r *http.Request) {
+		v := new(TemplateRepoRequest)
+		json.NewDecoder(r.Body).Decode(v)
+
+		testMethod(t, r, "POST")
```
Comment (gmlewis) [inline github/repos_test.go]: Please test for the required preview header here.

File: github/repos.go (hunk 4 above)
Comment (shanempope) [reply github/repos.go]: Thank you!
The docs show the header required throughout the endpoints, in order to get the fields. (Create for example)  https://developer.github.com/v3/repos/#create
I will add add it for all the required endpoints.

**Comment by shanempope** [comment]
I will add the custom headers to the other two endpoints later today.

**Comment by shanempope** [comment]
Alright, I manually tested all of the endpoints. It's ready for a second review. Thanks @gmlewis

**Comment by shanempope** [comment]
Thanks @gauntface and @gmlewis. I've squashed the commits

**Comment by gmlewis** [comment]
Thank you, @shanempope, but we prefer that you don't use force-push in this repo, as it makes it more difficult for other contributors to see the change history in case they wish to make comments during the PR review period.  Additionally, we always "Squash and merge" in this repo, so it ends up being a single commit anyway, so the force-push is unnecessary. 

Thank you, @gauntface.

Merging.

**Pull request merged**